### PR TITLE
Update icon generation

### PR DESCRIPTION
**Pull request by devoncarew**
- fix https://github.com/flutter/flutter-intellij/issues/1071 (show all material icons)
- fix https://github.com/flutter/flutter-intellij/issues/2711 (show images for cupertino icons)

This PR:
- commits a script to get the latest Material and Cupertino icon information from the flutter repo (from head)
- generates the necessary metadata files to enable generating the icons, and to enable IntelliJ to pull that info in at runtime
- commits a Flutter app which when run in the iOS simulator will generate 16px and 32px preview images for the material and cupertino icons

@jacob314 

This PR is split into a few different commits to facilitate review. Only the very brave should click on the last commit.

### CODE REVIEW

**Comment by devoncarew** [comment]
screen shot 2018-11-14 at 12 48 42 pm

screen shot 2018-11-14 at 1 17 01 pm

File: tool/icons/update_icons.dart [hunk 1]
```diff
@@ -1,105 +1,122 @@
...
+  HttpClientRequest request = await new HttpClient().getUrl(Uri.parse(url));
+  HttpClientResponse response = await request.close();
+  List<String> data = await response.transform(utf8.decoder).toList();
+  return data.join('');
+}
 
-  final String iconData = iconFile.readAsStringSync();
-  final String codepointData = codepointsFile.readAsStringSync();
-  final String newIconData = regenerateIconsFile(iconData, codepointData);
+// static const IconData threesixty = IconData(0xe577,
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply tool/icons/update_icons.dart]: This is meant to indicate what the regex below will match against; I'll add a comment to indicate that.

File: tool/icon_generator/.gitignore [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+**/doc/api/
+.dart_tool/
+.flutter-plugins
+.packages
+.pub-cache/
+.pub/
+build/
+
+# Android related
+**/android/**/gradle-wrapper.jar
```
Comment (jacob314) [inline tool/icon_generator/.gitignore]: nit: it isn't possible to just ignore entire directories?

File: tool/icon_generator/README.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# icon_generator
+
+A new Flutter project.
```
Comment (jacob314) [inline tool/icon_generator/README.md]: tweak or remove :)

File: tool/icon_generator/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+name: icon_generator
+description: A new Flutter project.
```
Comment (jacob314) [inline tool/icon_generator/pubspec.yaml]: tweak  description

File: tool/icons/update_icons.dart [hunk 5]
```diff
@@ -1,105 +1,122 @@
...
-      .map(getIconDeclaration)
-      .join();
+void generateDart(
+    List<Icon> icons, String filename, String prefix, String import) {
+//  final List<IconData> icons = [
+//    Icons.visibility
+//  ];
+
+  StringBuffer buf = StringBuffer();
+  buf.writeln('// Generated file - do not edit.');
```
Comment (jacob314) [inline tool/icons/update_icons.dart]: nit: using ''' here would make the generated code a little easier to edit if needed.

File: tool/icon_generator/.gitignore (hunk 2 above)
Comment (devoncarew) [reply tool/icon_generator/.gitignore]: this is straight from `flutter create`; happy to simplify it if we want to

File: tool/icon_generator/README.md (hunk 3 above)
Comment (devoncarew) [reply tool/icon_generator/README.md]: will do

File: tool/icon_generator/pubspec.yaml (hunk 4 above)
Comment (devoncarew) [reply tool/icon_generator/pubspec.yaml]: 👍

File: tool/icon_generator/lib/main.dart [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'cupertino.dart' as cupertino;
+import 'material.dart' as material;
+
+const String root =
+    '/Users/devoncarew/projects/intellij/flutter-intellij/resources/flutter';
```
Comment (jacob314) [inline tool/icon_generator/lib/main.dart]: make this a relative path if possible.

Comment (devoncarew) [reply tool/icon_generator/lib/main.dart]: Yeah, this is a bit of a wart. I'm not 100% sure how to get any path info into the Flutter app. I don't think the cwd for the iOS Simulator process is useful info here, but I'll double check.

File: tool/icon_generator/lib/main.dart [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+  final Iterable<Element> elements = finder.evaluate();
+  Element element = elements.first;
+
+  Future<ui.Image> imageFuture = _captureImage(element);
+
+  final ui.Image image = await imageFuture;
+  final ByteData bytes = await image.toByteData(format: ui.ImageByteFormat.png);
+
+  new File(path).writeAsBytesSync(bytes.buffer.asUint8List());
```
Comment (jacob314) [inline tool/icon_generator/lib/main.dart]: nit: might as well use the async writeAsBytes as the whole method is async.

File: tool/icon_generator/lib/main.dart [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+class MyIconApp extends StatelessWidget {
+  MyIconApp(this.materialIcons, this.cupertinoIcons)
+      : super(key: new UniqueKey());
+
+  final List<material.IconTuple> materialIcons; // 16x16
+  final List<cupertino.IconTuple> cupertinoIcons; // 32x32
+
+  @override
+  Widget build(BuildContext context) {
+    const Color color = const Color(0xFF777777);
```
Comment (jacob314) [inline tool/icon_generator/lib/main.dart]: comment that this color is just for debugging and won't impact the icons.

Comment (devoncarew) [reply tool/icon_generator/lib/main.dart]: This is the color that the icon's image is generated in; we want this color as it does well in both IntelliJ's light theme and in Darkula. I'll comment to that effect.

File: tool/icon_generator/lib/main.dart (hunk 6 above)
Comment (devoncarew) [reply tool/icon_generator/lib/main.dart]: cwd is `/` :(

File: tool/icon_generator/lib/main.dart (hunk 8 above)
Comment (devoncarew) [reply tool/icon_generator/lib/main.dart]: screen shot 2018-11-14 at 1 46 07 pm

File: tool/icon_generator/lib/main.dart (hunk 7 above)
Comment (devoncarew) [reply tool/icon_generator/lib/main.dart]: done

File: tool/icons/update_icons.dart (hunk 5 above)
Comment (devoncarew) [reply tool/icons/update_icons.dart]: done

**Comment by devoncarew** [comment]
Updated w/ changes from review comments

File: tool/icon_generator/lib/main.dart (hunk 8 above)
Comment (jacob314) [reply tool/icon_generator/lib/main.dart]: I wish it was more convenient to tint icons in IntelliJ.

File: tool/icon_generator/.gitignore (hunk 2 above)
Comment (jacob314) [reply tool/icon_generator/.gitignore]: no need.

**Comment by devoncarew** [comment]
bombs away

**Pull request merged**